Jasmine/Grid/ViewControllers/GridGameViewController.swift

```diff
@@ -0,0 +1,221 @@
+
+    /* Properties */
+    /// Stores a list of Chinese characters, which serves as the data source for  
+    /// `charactersCollectionView`.
+    fileprivate var chineseTexts: [String?]
+        = ["天", "翻", "地", "覆",
+           "CS32", "17", ":D", ":(",
+           "天", "翻", "地", "覆",
+           nil, "😛", nil, "😉"]
+
+    fileprivate var draggingTile: SquareTextViewCell?
+    fileprivate var draggingStartFrame: CGRect?
+    fileprivate var draggingStartIndex: IndexPath?
+
+    /* View Controller Lifecycles */
+    /// Readjusts layout (such as cell size) upon auto-rotate.
+    override func didRotate(from fromInterfaceOrientation: UIInterfaceOrientation) {
+        gridCollectionView.performBatchUpdates(gridCollectionView.reloadData,
+                                               completion: nil)
+    }
+
+    /* Gesture Recognisers */
+    /// Listens to a drag gesture and handles the operation of dragging a tile, and dropping it
+    /// to another location.
+    @IBAction func onTilesDragged(_ sender: UIPanGestureRecognizer) {
+        let position = sender.location(in: gridCollectionView)
+
+        switch sender.state {
+        case .began:
+            handleTileSelected(at: position)
+
+        case .changed:
+            handleTileDragged(at: position)
+
+        case .ended:
+            handleTileLanding(at: position)
+
+        default:
+            break
+        }
+    }
+}
+
+// MARK: - Drag and Drop Tiles
+fileprivate extension GridGameViewController {
+
+    /// Handles the case when the tile is lifted from the collection view.
+    /// Such a tile gets "detached" from the view by having its identity noted down in the 
+    /// properties.
+    ///
+    /// - Parameter position: location where the tile is selected.
+    fileprivate func handleTileSelected(at position: CGPoint) {
+        guard draggingTile == nil else {
+            return
+        }
+        guard draggingStartFrame == nil else {
+            return
+        }
+        guard let indexTouched = gridCollectionView.indexPathForItem(at: position) else {
+            return
+        }
+        guard let cellTouched = gridCollectionView
+            .cellForItem(at: indexTouched) as? SquareTextViewCell else {
+                return
+        }
+
+        gridCollectionView.bringSubview(toFront: cellTouched)
+        draggingTile = cellTouched
+        draggingStartFrame = cellTouched.frame
+        draggingStartIndex = indexTouched
+    }
+
+    /// Drags the tile that is referenced in `draggingTile`.
+    ///
+    /// - Parameter position: location where the tile should drag to.
+    fileprivate func handleTileDragged(at position: CGPoint) {
+        guard let draggingTile = draggingTile else {
+            return
+        }
+        let size = draggingTile.frame.size
+        let newOrigin = GeometryUtils.getOrigin(from: position, withSize: size)
+        draggingTile.frame.origin = newOrigin
```

Comment:
`frame` is a calculated property, use `bounds` instead. Also doesn't the origin automatically recalculates?